packages/vercel-ai-provider/src/edgedb-chat-language-model.ts

```diff
@@ -0,0 +1,574 @@
+    seed,
+    // providerMetadata: exists in the Vercel SDK d.ts but none of the providers use it
+  }: Parameters<LanguageModelV1["doGenerate"]>[0]) {
+    const type = mode.type;
+
+    const warnings: LanguageModelV1CallWarning[] = [];
+
+    if (frequencyPenalty != null) {
+      warnings.push({
+        type: "unsupported-setting",
+        setting: "frequencyPenalty",
+      });
+    }
+
+    if (presencePenalty != null) {
+      warnings.push({
+        type: "unsupported-setting",
+        setting: "presencePenalty",
+      });
+    }
+
+    if (stopSequences != null) {
+      warnings.push({
+        type: "unsupported-setting",
+        setting: "stopSequences",
+      });
+    }
+
+    if (isAnthropicModel(this.modelId) && seed != null) {
+      warnings.push({
+        type: "unsupported-setting",
+        setting: "seed",
+      });
+    }
+
+    if (!isAnthropicModel(this.modelId) && topK != null) {
+      warnings.push({
+        type: "unsupported-setting",
+        setting: "topK",
+      });
+    }
+
+    if (responseFormat != null && responseFormat.type !== "text") {
+      warnings.push({
+        type: "unsupported-setting",
+        setting: "responseFormat",
+        details: "JSON response format is not supported.",
+      });
+    }
+
+    // if (
+    //   responseFormat != null &&
+    //   responseFormat.type === "json"
+    //   // && responseFormat.schema != null
+    // ) {
+    //   warnings.push({
+    //     type: "unsupported-setting",
+    //     setting: "responseFormat",
+    //     details: "JSON response is not supported",
+    //     // details: "JSON response format schema is not supported",
+    //   });
+    // }
+
+    const baseArgs = {
+      model: this.modelId,
+      messages: convertToEdgeDBMessages(prompt),
+      temperature,
+      max_tokens: maxTokens,
+      top_p: topP,
+      seed,
+      // response_format: ...
+
+      // Mistral only
+      safe_prompt: this.settings.safePrompt,
+
+      // Anthropic only
+      top_k: topK,
+
+      // OpenAI only
+      echo: this.settings.echo,
+      logit_bias: this.settings.logitBias,
+      logprobs: this.settings.logprobs,
+      user: this.settings.user,
+    };
+
+    switch (type) {
+      case "regular": {
+        return {
+          args: {
+            ...baseArgs,
+            ...prepareTools(mode, this.modelId),
+          },
+          warnings,
+        };
+      }
+
+      // case 'object-json': {
+      // }
+
+      // case 'object-tool': {
+      // }
+
+      default: {
+        throw new Error(`Unsupported type: ${type}`);
+      }
+    }
+  }
+
+  async doGenerate(
+    options: Parameters<LanguageModelV1["doGenerate"]>[0],
+  ): Promise<Awaited<ReturnType<LanguageModelV1["doGenerate"]>>> {
+    const { args, warnings } = this.getArgs(options);
+    const { messages } = args;
+
+    const providedPromptId =
+      this.settings.prompt &&
+      ("name" in this.settings.prompt || "id" in this.settings.prompt);
```

Comment:
We seem to repeat this logic (and the `prompt` building logic below) a bunch. Can we export a utility that encapsulates this?

Reply:
we repeat this twice in the provider repo, in doGenerate and doStream, I can make a utility

Reply:
> providedPromptId is actually a very simple check, maybe we dont need a utility

Yeah, the "rule of three" says we should wait for one more case, right? 😅 My perspective on this "it's simple, just copy-paste" is fine if the logic is only a line or two, but the full logic of what this boolean represents and how it's used later is actually kind of complicated.

Mostly I care about a reviewer or someone coming along wanting to understand or change this code and the big spreading logic we do to construct the `prompt` object, they have to understand how this boolean is related to the later spread. A nicely named function can make this easier to understand without needing to dive too deeply into what it's doing unless that logic needs to change. And _if_ that logic needs to change, making it a shared function means it won't accidentally diverge in the two places.

Reply:
I created buildPrompt and buildQuery, I agree it is better. Thank you! Previously I tried to encapsulate in the separate function the whole body or even the whole object we pass to the `postJsonToApi` which then complicates types handling a bit more (especially for args and zod schemas), but is ofc also possible.